### PR TITLE
[FrameworkBundle][Config] Ignore exceptions thrown during reflection classes autoload

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/issues/32499 with PHP 7.3+
| License       | MIT
| Doc PR        | -

The behavior when an exception is thrown in a class loader changed in PHP 7.3 (cf https://3v4l.org/OQPk9). That means that the `throwOnRequiredClass` trick that is done in the parent class of these cache warmers (`AbstractPhpFileCacheWarmer`) does not work anymore with PHP7.3+.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
I'm not sure we need to work around since the new behavior has been reverted on <7.4
See #32504 
For 7.4, there are more places that need to be changed, and I'd like we really understand the best way to fox the root issue.

**Comment by nicolas-grekas** [comment]
Hmm, this is different, and discussed in https://bugs.php.net/74372
@nikic can we consider fixing the BC break by having the new behvior only for child classes of `Error` and let `Exception` behave as before?

**Comment by nikic** [comment]
@nicolas-grekas I don't think introducing an Error/Exception distinction here makes sense. If you want to ignore exceptions, please ignore them explicitly.

We could revert from 7.3 and introduce in a later version, but given that we're at 7.3.8 already and this change is explicitly called out in the migration guide, I'm not sure doing that is a good idea. You'll have to deal with it at some point in either case...

**Comment by nicolas-grekas** [comment]
@nikic thanks for the feedback, OK.

@fancyweb I think the patch should be only this one:
```diff
--- a/src/Symfony/Component/Config/Resource/ClassExistenceResource.php
+++ b/src/Symfony/Component/Config/Resource/ClassExistenceResource.php
@@ -76,7 +76,7 @@ class ClassExistenceResource implements SelfCheckingResourceInterface, \Serializ
 
             try {
                 $exists = class_exists($this->resource) || interface_exists($this->resource, false) || trait_exists($this->resource, false);
-            } catch (\ReflectionException $e) {
+            } catch (\Throwable $e) {
                 if (0 >= $timestamp) {
                     unset(self::$existsCache[1][$this->resource]);
                     throw $e;
```

The other places should not be changed (according to my current understanding at least.)

**Comment by nicolas-grekas** [comment]
OK, I missed the `} catch (\ReflectionException $e) {` in warmers, I understand how the issue propagates here. We should factorize this logic now, inside throwOnRequiredClass I believe.

**Comment by fancyweb** [comment]
@nicolas-grekas Just pushed all tests + a deduplication of the code. I don't think it was what you wanted so I'm wiating for your feedback 😁

**Comment by nicolas-grekas** [comment]
`ClassExistenceResource::isFresh()` is also affected + any direct user of `PhpArrayAdapter::throwOnRequiredClass()` would too.
The plan could be:
- factorize in `ClassExistenceResource`
- find a way to use `ClassExistenceResource` instead of `PhpArrayAdapter` in `FrameworkBundle`
- don't patch `PhpArrayAdapter` but deprecate its `throwOnRequiredClass()` in 4.4

**Comment by fancyweb** [comment]
@nicolas-grekas 

> factorize in ClassExistenceResource
Done.

> find a way to use ClassExistenceResource instead of PhpArrayAdapter in FrameworkBundle
`symfony/config` is a requirement of the `FrameworkBundle` so it's alright.

> don't patch PhpArrayAdapter but deprecate its throwOnRequiredClass() in 4.4
Actually this method was introduced directly as `@internal`, so I just removed it.

File: src/Symfony/Bundle/FrameworkBundle/CacheWarmer/AnnotationsCacheWarmer.php
Comment (fancyweb) [inline src/Symfony/Bundle/FrameworkBundle/CacheWarmer/AnnotationsCacheWarmer.php]: This is another fix / improvement. Moving this catch on every calls allows to not fail totally on first fail but just on each failing cases.

File: src/Symfony/Component/Config/Resource/ClassExistenceResource.php
Comment (fancyweb) [inline src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: All `\Exception` should be made `\Throwable` when merging to upper branches that supports PHP 7.

File: src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php]: we should keep this for now, to keep cross-component compat.
let's mark the method `@internal to be removed in Symfony 5.0` instead

File: src/Symfony/Component/Config/Resource/ClassExistenceResource.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: can't we get rid of this one, by doing the check below only when $previous === null?

File: src/Symfony/Component/Config/Resource/ClassExistenceResource.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: not the case here: we don't want to swallow Error - only Exception

File: src/Symfony/Component/Config/Resource/ClassExistenceResource.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: should we add $previous as 3rd arg?

File: src/Symfony/Component/Config/Resource/ClassExistenceResource.php
Comment (fancyweb) [reply src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: Yes we can.

However, splitting responsabilities in 2 methods looks better for me :

`_throwOnRequiredClass` is a dedicated method for the `ClassExistenceResource` class.

`throwOnRequiredClass` can be used from anywhere. Having a check on an unrelated static property in this method doesn't help to understand the system.

Waiting for your final decision.

File: src/Symfony/Component/Config/Resource/ClassExistenceResource.php
Comment (fancyweb) [reply src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: Yes, let's add it since it adds information!

File: src/Symfony/Component/Config/Resource/ClassExistenceResource.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: Then we need two clearly named function - an underscore is not enough.
It'd also we could get rid of the try/catch around the method in warmers, doable?

Comment (fancyweb) [reply src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: > an underscore is not enough.
`loadRequiredClass` (with a comment that it's an autoload function) & `throwOnRequiredClass` ?

> It'd also we could get rid of the try/catch around the method in warmers, doable?
Technically doable but that would mean `throwOnRequiredClass` wouldn't always throw the `\ReflectionException` (to silence it in the case it comes from a warmer) + it would need to rethrow passed previous exception. It gives this method two new added behaviors depending on the passed arguments while it's already kind of complicated. From my point of view, it's easier to understand with the try catch blocks in the warmer.

Comment (fancyweb) [reply src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: @nicolas-grekas Just repushed. I did not change the names as they are just the best I could find 😕 To make things clearer, I added detailed comments. IMHO, since there are internal methods, it should be OK. Otherwise, waiting for naming help here.

Comment (nicolas-grekas) [reply src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: the second method could be named `dontThrowOnRequiredClass`, and with that name, we would remove the try/catch boilerplate

Comment (fancyweb) [reply src/Symfony/Component/Config/Resource/ClassExistenceResource.php]: Looking at this again.

Just realized that it needs to throw because there might be something done in the `\ReflectionException $e` catch block and in this case we want to catch the thrown `\ReflectionException` by the `throwOnRequiredClass` to reapply the same logic (cf what I did in the `ClassExistenceResource` with the goto).

However, I was able to reduce the try/catch boilerplate by rethrowing the previous exception directly in `throwOnRequiredClass` and by reusing a common method for warmers.

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @fancyweb.

**Pull request merged**
